+pro #15870 OSGi headers for experimental Streams and HTTP modules

rkrzewski opened the pull request.

rkrzewski committed:
+pro #15870 OSGi headers for experimental Streams and HTTP modules

rkrzewski commented on project/AkkaBuild.scala:
```diff
@@ -1382,14 +1385,37 @@ object AkkaBuild extends Build {
+    val httpXml = exports(Seq("akka.http.scaladsl.marshallers.xml"),
+        imports = Seq(streamAndHttpImport("akka.stream.*"),
+            streamAndHttpImport("akka.http.*")))
+    // TODO Scala 2.11+ needs also versionedImport("scala.xml.*", "1.0", "1.1")
```
It would be great if someone having a clue about SBT could give me a hand here

rkrzewski commented on project/AkkaBuild.scala:
```diff
@@ -1382,14 +1385,37 @@ object AkkaBuild extends Build {
+            streamAndHttpImport("akka.http.*"),
+            streamAndHttpImport("akka.parboiled2.*")))
 
     val httpTestKit = exports(Seq("akka.http.scaladsl.testkit.*"))
```
Is this module expected to be used in OSGi runtime?

ktoso replied:
Yes, it's the same style of testkit that we provide for actors - a set of utilities to make testing simpler.

2beaucoup replied:
related: https://github.com/akka/akka/issues/17956 (javadsl is missing in artifacts)

rkrzewski replied:
If you really mean to support running the tests _inside_ an OSGi framework with PAX Exam or similar, testkit modules need OSGi headers too.

ktoso replied:
I'm not sure, we probably should - I mean what if someone uses stuff like arquillian etc. Does that count as osgi container? It may I gues..

ktoso commented:
refs #15870

ktoso commented:
OK TO TEST

rkrzewski commented:
I have amended the commit with OSGi settings for akka-stream-testkit and akka-http-testkit
akka-http-testkit is problematic because it depends on scalatest. scalatest is available as a bundle, but a proorly constructed one: it has mandatory dependencies on a ton of stuff (jmock, easymock, mockito, selenium, ant and more) and some of the dependencies is _not_ available as bundles, most notably ... junit and hamcrest.
Does it make sense to use akka-http-testkit without scalatest in the runtime classpath? Making the dependency optional would give the bundle a chance to be loaded.

rkrzewski commented:
merge conflict, will rebase.

rkrzewski commented:
There's one remaining problem: `scala.xml` needs to be imported with different version when building against Scala 2.11+. I'd like do do something like this:
```scala
def scalaXmlImport: Seq[String] = 
  if(scalaVersion == "2.10") Nil
  else Seq(versionedImport("scala.xml.*", "1.0", "1.1"))

val httpXml = exports(Seq("akka.http.scaladsl.marshallers.xml"),
        imports = Seq(streamAndHttpImport("akka.stream.*"),
            streamAndHttpImport("akka.http.*")) ++ scalaXmlImport)
```
but I don't know the proper SBT idiom to do the ```scalaVersion == "2.10"``` part.

rkrzewski committed:
+pro #15870 generate correct version range for scala.xml import on Scala 2.11+

scala.xml was extracted into a separate module with independent relase cycle
starting with Scala 2.11.0. In order to generate correct OSGi bundles for
this environment, an import for scala.xml.* with appropriate version range
must be inserted before scala.* import in BND instructions.

rkrzewski commented:
Done. I verified locally that correct imports are generated both in 2.10 and 2.11 builds.
If you prefer those 2 commits squashed, let me know.

2beaucoup commented on project/AkkaBuild.scala:
```diff
@@ -1382,16 +1385,43 @@ object AkkaBuild extends Build {
+            streamAndHttpImport("akka.parboiled2.*")))
 
-    val httpCore = Nil // FIXME #15689
+    val httpTestKit = exports(Seq("akka.http.scaladsl.testkit.*"),
```
javadsl?

rkrzewski replied:
fixed

ktoso commented on project/AkkaBuild.scala:
```diff
@@ -1428,11 +1458,17 @@ object AkkaBuild extends Build {
     def exports(packages: Seq[String] = Seq(), imports: Seq[String] = Nil) = osgiSettings ++ Seq(
-      OsgiKeys.importPackage := imports ++ defaultImports,
+      OsgiKeys.importPackage := imports,
+      OsgiKeys.importPackage <++= scalaVersion { v =>
```
`"2.10"` is `scalaBinaryVersion` :)

jrudolph replied:
Are you sure this change is at the right place? This will add `scala.xml` to any package that uses `exports`, right?

rkrzewski replied:
@jrudolph no worries, BND is smart and will add it to `Import-Package` header only when actually needed. I do this in this place, because it needs to be in front of `scala.*` to be effective.

@ktoso should I change the condition to `scalaBinaryVersion { v => if(v == "2.10") ...` or should it stay as is? I actually nicked that from cpsPlugin settings declaration at line 170

ktoso commented:
Looks great, thank you @rkrzewski!
After merging all other PRs I'll get to this one and verify we're not missing anything, then - release 1.0!

2m commented:
I wrote a [quick script](https://gist.github.com/2m/51fc0490ffea7bdc4cb9) to verify that jars contain all of the compiled class files. So this LGTM.

ktoso commented:
thanks! That's super helpful, so confidently merging this.

ktoso commented on project/AkkaBuild.scala:
```diff
@@ -1382,16 +1385,43 @@ object AkkaBuild extends Build {
+            streamAndHttpImport("akka.parboiled2.*")))
 
-    val httpCore = Nil // FIXME #15689
+    val httpTestKit = exports(Seq("akka.http.scaladsl.testkit.*", "akka.http.javaadsl.testkit.*"),
```
excellent, thanks @rkrzewski

jrudolph replied:
Bam! `javaadsl` spoiled the show. Hard to spot, though...

ktoso replied:
aaaaargh :<

rkrzewski replied:
sorry about that :(

jrudolph replied:
No worries about that, this is so easy to get wrong. The problem is that there are no safeguards that would catch a problem like this.

ktoso replied:
Yeah, no worries - I mean, I reviewed and looked at this exact line and still missed the typo.
Will be fixed soon in 1.1 tho

rkrzewski replied:
Hmm, shouldn't the script written by @2m have detected this problem?

Also, the verification script should check that the generated jar does not contain _more_ classes than `target/classes` folder, because it's easy to accidentally re-export packages from an upstream dependency if a too broad Export declaration is used (which is completely valid, and occasionally useful in OSGi environment)

rkrzewski commented:
:sunglasses:
